packages/drivers/odsp-driver/src/getFileLink.ts

```diff
@@ -111,11 +117,27 @@ async function getFileLinkCore(
                 };
                 const response = await fetchHelper(url, requestInit);
                 additionalProps = response.commonSpoHeaders;
-                if (response.content.ok) {
-                    const sharingInfo = await response.content.json();
-                    return sharingInfo?.d?.directUrl as string;
+
+                // We actually already know the response is ok, or else the fetchHelper would have thrown
+                // (it does an even more thorough check).  Keeping it here though for layers of protection.
+                if (!response.content.ok) {
+                    throwOdspNetworkError(
+                        `odspFetchError [${response.content.status}]`,
+                        response.content.status,
+                        response.content,
+                        await response.content.text(),
+                    );
                 }
-                return undefined;
+                const sharingInfo = await response.content.json();
+                const directUrl = sharingInfo?.d?.directUrl;
+                if (typeof directUrl !== "string") {
+                    // Here use a generic error.  fetchIncorrectResponse might be preferable but will result in a
+                    // retry, which we don't want.  Both because we don't have a reason to expect a different
+                    // response on retry, and also because the retry loop in in getFileLink doesn't have the same
+                    // "retry once" logic that getWithRetryForTokenRefresh does.
+                    throwOdspNetworkError("malformedGetSharingInformationResponse", 400);
```

Comment:
Would particularly appreciate feedback on appropriate error here.  I started with `fetchIncorrectResponse` which seems more correct, but it hits an issue that we have a double-loop of retry logic here (`getWithRetryForTokenRefresh` and then the do-while up in `getFileLink`).  Only `getWithRetryForTokenRefresh` knows how to do a "retry once" currently, so the do-while loops infinitely.  We could prob solve with a more thorough rewrite.

Reply:
What about `incorrectServerResponse`?  From DriverErrorType:
```
    /**
     * Unexpected response from server. Either JSON is malformed, or some required properties are missing
     */
    incorrectServerResponse = "incorrectServerResponse",
```

Reply:
In a previous iteration I threw a custom `NonRetryableError` using `DriverErrorType.incorrectServerResponse`, but debated whether it made things more confusing to have that error become a "sometimes-retried" error.  It's an option though.

Reply:
fetchIncorrectResponse is the correct code to use here. It will be mapped to DriverErrorType.incorrectServerResponse which will instruct getWithRetryForTokenRefresh to retry, but only once.
The issue with infinite loop that that Matt described seems like a bug - this error should not be retryable, i.e. code in createOdspNetworkError seems wrong. Looks like I made a mistake - see https://github.com/microsoft/FluidFramework/commit/e9d70c9f5fa7a2e0661c2b7c95cbe4e8bf7c6020#diff-b33acf473b848a8cf608555fb60fec5cfb7bd3e395c91b6e1bb90ffa084dd68c

Reply:
Latest commit changes `fetchIncorrectResponse` to a `NonRetryableError` and uses it.